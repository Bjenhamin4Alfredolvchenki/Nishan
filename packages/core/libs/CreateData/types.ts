import { ICache } from '@nishans/cache';
import { IOperation, TTextFormat } from '@nishans/types';
import { Logger } from '../../types';

export interface ParentCollectionData {
	parent_collection_id: string;
	name: TTextFormat;
	token: string;
	logger?: Logger;
	stack: IOperation[];
	cache: ICache;
	parent_relation_schema_unit_id: string;
}
