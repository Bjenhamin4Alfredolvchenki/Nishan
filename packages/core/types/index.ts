export * from './block';
export * from './maps';
export * from './nishan';
