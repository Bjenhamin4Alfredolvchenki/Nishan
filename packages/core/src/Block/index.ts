import CollectionViewPage from './CollectionViewPage';
import CollectionView from './CollectionView';
import Page from './Page';
import CollectionBlock from './CollectionBlock';
import Block from './Block';

export { Block, CollectionViewPage, CollectionView, Page, CollectionBlock };
