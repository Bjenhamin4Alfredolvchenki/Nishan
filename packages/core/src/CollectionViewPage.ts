import { ISpace, IPage, ICollectionViewPage } from '@nishans/types';
import { NotionPermissions } from '../src';
import { NishanArg } from '../types';
import CollectionBlock from './CollectionBlock';

/**
 * A class to represent collectionviewpage of Notion
 * @noInheritDoc
 */
class CollectionViewPage extends CollectionBlock<ICollectionViewPage> {
	Permissions: NotionPermissions;
	constructor (arg: NishanArg) {
		super({ ...arg });
		this.Permissions = new NotionPermissions(arg, arg.id, 'block');
	}

	getCachedParentData () {
		const data = this.getCachedData();
		return this.cache[data.parent_table].get(data.parent_id) as IPage | ISpace;
	}
}

export default CollectionViewPage;
