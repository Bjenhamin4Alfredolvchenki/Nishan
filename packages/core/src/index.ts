import {NotionCacheClass} from "@nishans/cache";
import { INotionUser } from "@nishans/types";

import Block from "./Block";
import Collection from "./Collection";
import CollectionViewPage from "./CollectionViewPage";
import CollectionView from "./CollectionView";
import NotionUser from "./NotionUser";
import Page from "./Page";
import UserSettings from "./UserSettings";
import UserRoot from "./UserRoot";
import SpaceView from "./SpaceView";
import Space from "./Space";
import SchemaUnit from "./SchemaUnit";
import CollectionBlock from "./CollectionBlock";
import NotionPermissions from "./Permissions";
export * from "./View";
import { constructLogger, error, iterateAndGetChildren, transformToMultiple } from "../utils";
import {Logger, NishanArg,FilterType, FilterTypes} from "../types";

class Nishan extends NotionCacheClass {
  token: string;
  interval: number;
  init_cache: boolean;
  logger: Logger;

  constructor(arg: Pick<NishanArg, "token"> & { interval?: number, logger?: Logger }) {
    super(arg);
    this.token = arg.token;
    this.interval = arg.interval ?? 500;
    this.init_cache = false;
    this.logger = constructLogger(arg.logger)
  }

  #initializeCache = async () => {
    if (!this.init_cache) {
      await this.initializeNotionCache();
      this.init_cache = true;
    }
  }

  /**
   * Get `INotionUser` and return `NotionUser` by the `args` param
   * @param args An string id, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUser(arg?: FilterType<INotionUser>) {
    return (await this.getNotionUsers(transformToMultiple(arg), false))[0];
  }

  /**
   * Get `INotionUser[]` and return `NotionUser[]` by the `args` param
   * @param args An array of string ids, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUsers(args?: FilterTypes<INotionUser>, multiple?: boolean) {
    await this.#initializeCache();

    const common_props = {
      token: this.token,
      cache: this.cache,
      interval: this.interval,
      logger: this.logger,
      stack: [],
    }

    const notion_user_ids: string[] = [];
    for (const [notion_user_id] of this.cache.notion_user) {
      notion_user_ids.push(notion_user_id)
    }

    return await iterateAndGetChildren<INotionUser, INotionUser, NotionUser[]>(args, (id)=>this.cache.notion_user.get(id), {
      ...common_props,
      multiple,
      child_ids: notion_user_ids,
      child_type: 'notion_user',
      container: [],
      parent_id: '',
      parent_type: 'notion_user'
    }, (id, _, container)=>container.push(new NotionUser({ ...common_props, user_id: id, id, space_id: "0", shard_id: 0 })));
  }
}

export default Nishan;
export * from "../types";
export * from "../utils";

export {
  Block,
  Collection,
  CollectionViewPage,
  CollectionView,
  NotionUser,
  Page,
  UserSettings,
  UserRoot,
  SpaceView,
  Space,
  SchemaUnit,
  CollectionBlock,
  NotionPermissions
}