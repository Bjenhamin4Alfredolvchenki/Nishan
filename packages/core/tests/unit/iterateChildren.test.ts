import { ICache } from '@nishans/endpoints';
import { IOperation, IPage, ISpace, TBlock, TPage } from '@nishans/types';
import deepEqual from 'deep-equal';
import { iterateAndDeleteChildren, iterateAndGetChildren, iterateChildren } from '../../src';
import colors from 'colors';

describe('iterateChildren', () => {
	describe('Array of ids', () => {
		it('Multiple false', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container: string[] = [];
			await iterateChildren<TPage>(child_ids, (id) => container.push(id), (id) => cache.block.get(id) as TPage, {
				parent_id: 'parent_one_id',
				parent_type: 'block',
				child_type: 'block',
				child_ids,
				multiple: false
			});

			expect(deepEqual(container, [ 'child_one_id' ]));
		});

		it('Shows warning if child doesnt exist in the cache', async () => {
			const child_ids = [ 'child_one_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					]
				])
			} as any;

			const console_log_spy = jest.spyOn(console, 'log');

			await iterateChildren<TPage>(
				child_ids,
				() => {
					return;
				},
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids
				}
			);

			expect(console_log_spy).toHaveBeenCalledTimes(1);
			expect(console_log_spy).toHaveBeenCalledWith(
				colors.yellow.bold('block:child_one_id does not exist in the cache')
			);
			console_log_spy.mockClear();
		});

		it('Shows warning if child_id is not present in child_ids', async () => {
			const child_ids = [ 'child_one_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					]
				])
			} as any;

			const console_log_spy = jest.spyOn(console, 'log');
			await iterateChildren<TPage>(
				child_ids,
				() => {
					return;
				},
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids: []
				}
			);

			expect(console_log_spy).toHaveBeenCalledTimes(1);
			expect(console_log_spy).toHaveBeenCalledWith(
				colors.yellow.bold('block:child_one_id is not a child of block:parent_one_id')
			);
			console_log_spy.mockClear();
		});
	});

	describe('Callback', () => {
		it('Multiple false', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container: string[] = [];
			await iterateChildren<TPage>(() => true, (id) => container.push(id), (id) => cache.block.get(id) as TPage, {
				parent_id: 'parent_one_id',
				parent_type: 'block',
				child_type: 'block',
				child_ids,
				multiple: false
			});

			expect(deepEqual(container, [ 'child_one_id' ]));
		});

		it('Shows warning if child doesnt exist in the cache', async () => {
			const child_ids = [ 'child_one_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					]
				])
			} as any;

			const console_log_spy = jest.spyOn(console, 'log');

			await iterateChildren<TPage>(
				() => true,
				() => {
					return;
				},
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids
				}
			);

			expect(console_log_spy).toHaveBeenCalledTimes(1);
			expect(console_log_spy).toHaveBeenCalledWith(
				colors.yellow.bold('block:child_one_id does not exist in the cache')
			);
			console_log_spy.mockClear();
		});
	});
});

describe('iterateAndGetChildren', () => {
	describe('Array of ids', () => {
		it('Return container item', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container = await iterateAndGetChildren<ISpace, TPage, TBlock[]>(
				child_ids,
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids,
					container: [],
					cache,
					logger () {
						return;
					}
				},
				(_, data, container) => container.push(data)
			);

			expect(
				deepEqual(container, [
					{
						id: 'child_one_id'
					},
					{
						id: 'child_two_id'
					}
				])
			);
		});

		it('Works for string child_id path', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container = await iterateAndGetChildren<IPage, TBlock, TBlock[]>(
				child_ids,
				(id) => cache.block.get(id) as TBlock,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids: 'content',
					container: [],
					cache
				},
				(_, data, container) => container.push(data)
			);

			expect(
				deepEqual(container, [
					{
						id: 'child_one_id'
					},
					{
						id: 'child_two_id'
					}
				])
			);
		});
	});

	describe('Callbacks', () => {
		it('Return container item', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container = await iterateAndGetChildren<ISpace, TPage, TBlock[]>(
				() => true,
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_ids,
					cache,
					child_type: 'block',
					container: []
				},
				(_, data, container) => container.push(data)
			);

			expect(
				deepEqual(container, [
					{
						id: 'child_one_id'
					},
					{
						id: 'child_two_id'
					}
				])
			);
		});

		it('undefined args', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container = await iterateAndGetChildren<ISpace, TPage, TBlock[]>(
				undefined,
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids,
					container: [],
					cache,
					multiple: false
				},
				(_, data, container) => container.push(data)
			);

			expect(
				deepEqual(container, [
					{
						id: 'child_one_id'
					}
				])
			);
		});

		it('false args', async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ];

			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							contents: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const container = await iterateAndGetChildren<ISpace, TPage, TBlock[]>(
				() => false,
				(id) => cache.block.get(id) as TPage,
				{
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_type: 'block',
					child_ids,
					container: [],
					cache,
					multiple: false
				},
				(_, data, container) => container.push(data)
			);

			expect(deepEqual(container, []));
		});
	});
});

describe('iterateAndDeleteChildren', () => {
	describe('Array of ids', () => {
		it(`Array child_ids`, async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ],
				stack: IOperation[] = [];
			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							type: 'page',
							content: [ 'child_one_id', 'child_two_id', 'child_three_id' ]
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					],
					[
						'child_three_id',
						{
							id: 'child_three_id'
						}
					]
				])
			} as any;

			const deleted_data = await iterateAndDeleteChildren<IPage, TBlock>(
				[ 'child_one_id', 'child_two_id' ],
				(id) => cache.block.get(id),
				{
					container: [],
					cache,
					child_ids,
					child_type: 'block',
					parent_id: 'parent_one_id',
					parent_type: 'block',
					child_path: 'content',
					stack,
					user_id: 'user_id',
					logger: (method, child_type, child_id) => {
						expect(method).toBe('DELETE');
						expect(child_type).toBe('block');
						expect(child_id).toBeTruthy();
					}
				},
				(id, data) => {
					expect(id).toBeTruthy();
					expect(data).toBeTruthy();
				}
			);
			const parent_data = cache.block.get('parent_one_id') as IPage;

			expect(parent_data).toMatchSnapshot({
				content: [ 'child_three_id' ],
				last_edited_by_id: 'user_id',
				last_edited_by_table: 'notion_user',
				last_edited_time: expect.any(Number)
			});

			expect(deleted_data).toMatchSnapshot([
				{
					id: 'child_one_id',
					alive: false,
					last_edited_by_table: 'notion_user',
					last_edited_by_id: 'user_id',
					last_edited_time: expect.any(Number)
				},
				{
					id: 'child_two_id',
					alive: false,
					last_edited_by_table: 'notion_user',
					last_edited_by_id: 'user_id',
					last_edited_time: expect.any(Number)
				}
			]);

			expect(stack).toMatchSnapshot([
				{
					command: 'update',
					table: 'block',
					id: 'child_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				},
				{
					command: 'listRemove',
					table: 'block',
					id: 'parent_one_id',
					path: [ 'content' ],
					args: {
						id: 'child_one_id'
					}
				},
				{
					command: 'update',
					table: 'block',
					id: 'child_two_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				},
				{
					command: 'listRemove',
					table: 'block',
					id: 'parent_one_id',
					path: [ 'content' ],
					args: {
						id: 'child_two_id'
					}
				},
				{
					command: 'update',
					table: 'block',
					id: 'parent_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				}
			]);
		});

		it(`String child_ids`, async () => {
			const child_ids = [ 'child_one_id', 'child_two_id', 'child_three_id' ],
				stack: IOperation[] = [];
			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							type: 'page',
							content: [ 'child_one_id', 'child_two_id' ]
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const deleted_data = await iterateAndDeleteChildren<IPage, TBlock>(
				[ 'child_one_id' ],
				(id) => cache.block.get(id),
				{
					cache,
					container: [],
					child_path: 'content',
					child_ids: 'content',
					child_type: 'block',
					parent_id: 'parent_one_id',
					parent_type: 'block',
					stack,
					user_id: 'user_id'
				}
			);
			const parent_data = cache.block.get('parent_one_id') as IPage;
			expect(parent_data).toMatchSnapshot({
				content: [ 'child_two_id' ],
				last_edited_by_id: 'user_id',
				last_edited_by_table: 'notion_user',
				last_edited_time: expect.any(Number)
			});

			expect(deleted_data).toMatchSnapshot([
				{
					id: 'child_one_id',
					alive: false,
					last_edited_by_table: 'notion_user',
					last_edited_by_id: 'user_id',
					last_edited_time: expect.any(Number)
				}
			]);

			expect(stack).toMatchSnapshot([
				{
					command: 'update',
					table: 'block',
					id: 'child_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				},
				{
					command: 'listRemove',
					table: 'block',
					id: 'parent_one_id',
					path: [ 'content' ],
					args: {
						id: 'child_one_id'
					}
				},
				{
					command: 'update',
					table: 'block',
					id: 'parent_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				}
			]);
		});

		it(`Multiple false`, async () => {
			const child_ids = [ 'child_one_id', 'child_two_id', 'child_three_id' ],
				stack: IOperation[] = [];
			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							type: 'page',
							content: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					],
					[
						'child_three_id',
						{
							id: 'child_three_id'
						}
					]
				])
			} as any;
			const deleted_data = await iterateAndDeleteChildren<IPage, TBlock>(
				[ 'child_one_id', 'child_two_id' ],
				(id) => cache.block.get(id),
				{
					container: [],
					child_path: 'content',
					cache,
					child_ids: 'content',
					multiple: false,
					child_type: 'block',
					parent_id: 'parent_one_id',
					parent_type: 'block',
					stack,
					user_id: 'user_id'
				}
			);
			const parent_data = cache.block.get('parent_one_id') as IPage;
			expect(parent_data as any).toMatchSnapshot({
				id: 'parent_one_id',
				content: [ 'child_two_id', 'child_three_id' ],
				last_edited_by_table: 'notion_user',
				last_edited_by_id: 'user_id',
				last_edited_time: expect.any(Number)
			});
			expect(deleted_data as any).toMatchSnapshot([
				{
					id: 'child_one_id',
					alive: false,
					last_edited_by_table: 'notion_user',
					last_edited_by_id: 'user_id',
					last_edited_time: expect.any(Number)
				}
			]);
			expect(stack).toMatchSnapshot([
				{
					args: {
						alive: false,
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					},
					command: 'update',
					id: 'child_one_id',
					path: [],
					table: 'block'
				},
				{
					command: 'listRemove',
					table: 'block',
					id: 'parent_one_id',
					path: [ 'content' ],
					args: {
						id: 'child_one_id'
					}
				},
				{
					command: 'update',
					table: 'block',
					id: 'parent_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				}
			]);
		});

		it(`parent doesnt contain child`, async () => {
			const stack: IOperation[] = [];
			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id'
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					]
				])
			} as any;
			const deleted_data = await iterateAndDeleteChildren<IPage, TBlock>([], (id) => cache.block.get(id), {
				cache,
				child_ids: 'content',
				multiple: false,
				child_type: 'block',
				parent_id: 'parent_one_id',
				parent_type: 'block',
				stack,
				user_id: 'user_id',
				container: [],
				child_path: 'content'
			});
			const parent_data = cache.block.get('parent_one_id') as IPage;
			expect(
				deepEqual(parent_data, {
					id: 'parent_one_id'
				})
			).toBe(true);
			expect(deepEqual(deleted_data, [])).toBe(true);
			expect(deepEqual(stack, [])).toBe(true);
		});

		it(`manual true`, async () => {
			const stack: IOperation[] = [];
			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							content: [ 'child_one_id' ]
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;

			const deleted_data = await iterateAndDeleteChildren<IPage, TBlock>(
				[ 'child_one_id' ],
				(id) => cache.block.get(id),
				{
					cache,
					container: [],
					child_ids: 'content',
					child_type: 'block',
					parent_id: 'parent_one_id',
					parent_type: 'block',
					stack,
					manual: true,
					user_id: 'user_id'
				},
				(_, child_data) => (child_data.alive = false)
			);
			const parent_data = cache.block.get('parent_one_id') as IPage;

			expect(parent_data as any).toMatchSnapshot({
				id: 'parent_one_id',
				last_edited_by_table: 'notion_user',
				last_edited_by_id: 'user_id',
				last_edited_time: expect.any(Number),
				content: [ 'child_one_id' ]
			});

			expect(
				deepEqual(deleted_data, [
					{
						id: 'child_one_id',
						alive: false
					}
				])
			).toBe(true);

			expect(stack).toMatchSnapshot([
				{
					command: 'update',
					table: 'block',
					id: 'parent_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				}
			]);
		});

		it(`manual false, child_path undefined`, async () => {
			const child_ids = [ 'child_one_id', 'child_two_id' ],
				stack: IOperation[] = [];
			const cache: ICache = {
				block: new Map([
					[
						'parent_one_id',
						{
							id: 'parent_one_id',
							type: 'page',
							content: child_ids
						}
					],
					[
						'child_one_id',
						{
							id: 'child_one_id'
						}
					],
					[
						'child_two_id',
						{
							id: 'child_two_id'
						}
					]
				])
			} as any;
			const deleted_data = await iterateAndDeleteChildren<IPage, TBlock>(
				[ 'child_one_id' ],
				(id) => cache.block.get(id),
				{
					container: [],
					cache,
					child_ids: 'content',
					multiple: false,
					child_type: 'block',
					parent_id: 'parent_one_id',
					parent_type: 'block',
					stack,
					user_id: 'user_id'
				} as any
			);
			const parent_data = cache.block.get('parent_one_id') as IPage;

			expect(parent_data as any).toMatchSnapshot({
				id: 'parent_one_id',
				content: [ 'child_one_id', 'child_two_id' ],
				last_edited_by_table: 'notion_user',
				last_edited_by_id: 'user_id',
				last_edited_time: expect.any(Number)
			});

			expect(deleted_data as any).toMatchSnapshot([
				{
					id: 'child_one_id',
					alive: false,
					last_edited_by_table: 'notion_user',
					last_edited_by_id: 'user_id',
					last_edited_time: expect.any(Number)
				}
			]);

			expect(stack).toMatchSnapshot([
				{
					args: {
						alive: false,
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					},
					command: 'update',
					id: 'child_one_id',
					path: [],
					table: 'block'
				},
				{
					command: 'update',
					table: 'block',
					id: 'parent_one_id',
					path: [],
					args: {
						last_edited_by_table: 'notion_user',
						last_edited_by_id: 'user_id',
						last_edited_time: expect.any(Number)
					}
				}
			]);
		});
	});
});
