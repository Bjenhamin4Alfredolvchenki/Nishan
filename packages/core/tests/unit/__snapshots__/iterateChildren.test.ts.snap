// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iterateAndDeleteChildren Array of ids Array child_ids 1`] = `
Object {
  "content": Array [
    "child_three_id",
  ],
  "id": "parent_one_id",
  "last_edited_by_id": "user_id",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "type": "page",
}
`;

exports[`iterateAndDeleteChildren Array of ids Array child_ids 2`] = `
Array [
  Object {
    "alive": false,
    "id": "child_one_id",
    "last_edited_by_id": "user_id",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
  },
  Object {
    "alive": false,
    "id": "child_two_id",
    "last_edited_by_id": "user_id",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids Array child_ids 3`] = `
Array [
  Object {
    "args": Object {
      "alive": false,
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "child_one_id",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "id": "child_one_id",
    },
    "command": "listRemove",
    "id": "parent_one_id",
    "path": Array [
      "content",
    ],
    "table": "block",
  },
  Object {
    "args": Object {
      "alive": false,
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "child_two_id",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "id": "child_two_id",
    },
    "command": "listRemove",
    "id": "parent_one_id",
    "path": Array [
      "content",
    ],
    "table": "block",
  },
  Object {
    "args": Object {
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "parent_one_id",
    "path": Array [],
    "table": "block",
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids Multiple false 1`] = `
Object {
  "content": Array [
    "child_two_id",
    "child_three_id",
  ],
  "id": "parent_one_id",
  "last_edited_by_id": "user_id",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "type": "page",
}
`;

exports[`iterateAndDeleteChildren Array of ids Multiple false 2`] = `
Array [
  Object {
    "alive": false,
    "id": "child_one_id",
    "last_edited_by_id": "user_id",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids Multiple false 3`] = `
Array [
  Object {
    "args": Object {
      "alive": false,
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "child_one_id",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "id": "child_one_id",
    },
    "command": "listRemove",
    "id": "parent_one_id",
    "path": Array [
      "content",
    ],
    "table": "block",
  },
  Object {
    "args": Object {
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "parent_one_id",
    "path": Array [],
    "table": "block",
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids String child_ids 1`] = `
Object {
  "content": Array [
    "child_two_id",
  ],
  "id": "parent_one_id",
  "last_edited_by_id": "user_id",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "type": "page",
}
`;

exports[`iterateAndDeleteChildren Array of ids String child_ids 2`] = `
Array [
  Object {
    "alive": false,
    "id": "child_one_id",
    "last_edited_by_id": "user_id",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids String child_ids 3`] = `
Array [
  Object {
    "args": Object {
      "alive": false,
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "child_one_id",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "id": "child_one_id",
    },
    "command": "listRemove",
    "id": "parent_one_id",
    "path": Array [
      "content",
    ],
    "table": "block",
  },
  Object {
    "args": Object {
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "parent_one_id",
    "path": Array [],
    "table": "block",
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids manual false, child_path undefined 1`] = `
Object {
  "content": Array [
    "child_one_id",
    "child_two_id",
  ],
  "id": "parent_one_id",
  "last_edited_by_id": "user_id",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "type": "page",
}
`;

exports[`iterateAndDeleteChildren Array of ids manual false, child_path undefined 2`] = `
Array [
  Object {
    "alive": false,
    "id": "child_one_id",
    "last_edited_by_id": "user_id",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids manual false, child_path undefined 3`] = `
Array [
  Object {
    "args": Object {
      "alive": false,
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "child_one_id",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "parent_one_id",
    "path": Array [],
    "table": "block",
  },
]
`;

exports[`iterateAndDeleteChildren Array of ids manual true 1`] = `
Object {
  "content": Array [
    "child_one_id",
  ],
  "id": "parent_one_id",
  "last_edited_by_id": "user_id",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
}
`;

exports[`iterateAndDeleteChildren Array of ids manual true 2`] = `
Array [
  Object {
    "args": Object {
      "last_edited_by_id": "user_id",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "parent_one_id",
    "path": Array [],
    "table": "block",
  },
]
`;
